MREUnityRuntime/MREUnityRuntimeLib/Animation/Animation.cs

```diff
@@ -91,88 +120,49 @@ internal override void Update()
-			 * Process tracks
-			 *********************************************/
-
-			JToken jToken;
-			JValue jValue = new JValue(false);
-			JObject jObject = new JObject();
-
+			// process tracks
 			for (var ti = 0; ti < Data.Tracks.Length; ti++)
 			{
 				var track = Data.Tracks[ti];
+				(Keyframe prevFrame, Keyframe nextFrame) = GetActiveKeyframes(ti, currentTime);
 
-				// grab the same keyframes from the last update
-				Keyframe prevFrame = track.Keyframes[LastKeyframeIndex[ti]],
-					nextFrame = LastKeyframeIndex[ti] + 1 < Data.Tracks.Length ? track.Keyframes[LastKeyframeIndex[ti] + 1] : null;
-
-				// if the current time isn't in that range, try the "next" keyframe based on speed sign
-				if (currentTime < prevFrame.Time || currentTime >= nextFrame.Time)
-				{
-					if (Speed < 0 && LastKeyframeIndex[ti] > 0)
-					{
-						nextFrame = prevFrame;
-						prevFrame = track.Keyframes[LastKeyframeIndex[ti] - 1];
-						LastKeyframeIndex[ti]--;
-					}
-					else if (Speed > 0 && LastKeyframeIndex[ti] < track.Keyframes.Length - 2)
-					{
-						prevFrame = nextFrame;
-						nextFrame = track.Keyframes[LastKeyframeIndex[ti] + 1];
-						LastKeyframeIndex[ti]++;
-					}
-				}
-
-				// if it's still not in range, we just have to search
-				if (prevFrame.Time > currentTime || nextFrame.Time <= currentTime)
+				// either no keyframes, or time out of range
+				if (prevFrame == null)
 				{
-					prevFrame = null;
-					nextFrame = null;
-					for (int i = 0; i < (track.Keyframes.Length - 1); i++)
-					{
-						if (currentTime >= track.Keyframes[i].Time && currentTime < track.Keyframes[i + 1].Time)
-						{
-							prevFrame = track.Keyframes[i];
-							nextFrame = track.Keyframes[i + 1];
-							LastKeyframeIndex[ti] = i;
-							break;
-						}
-					}
-					// either no keyframes, or time out of range
-					if (prevFrame == null)
-					{
-						continue;
-					}
+					continue;
 				}
-				
 
 				var linearT = (currentTime - prevFrame.Time) / (nextFrame.Time - prevFrame.Time);
 
 				// compute new value for targeted field
-				if (prevFrame.Value.Type == JTokenType.Object)
-				{
-					jToken = jObject;
-					Interpolations.Interpolate(prevFrame.Value, nextFrame.Value, linearT, ref jToken, nextFrame.Bezier ?? track.Bezier);
-				}
-				else
+				var interpolatedToken = TokenPool.Lease(prevFrame.Value.Type);
+				Interpolations.Interpolate(prevFrame.Value, nextFrame.Value, linearT, ref interpolatedToken, nextFrame.Bezier ?? track.Bezier);
+
+				// mix starting values with relative keyframe values
+				JToken mixedToken = interpolatedToken;
+				if (track.Relative == true)
 				{
-					jToken = jValue;
-					Interpolations.Interpolate(prevFrame.Value, nextFrame.Value, linearT, ref jToken, nextFrame.Bezier ?? track.Bezier);
+					mixedToken = TokenPool.Lease(interpolatedToken.Type);
+					Interpolations.ResolveRelativeValue(ImplicitStartKeyframes[ti].Value, interpolatedToken, ref mixedToken);
```

Comment:
I think you can preprocess the animation so that this doesn't need to be called every frame.

Reply:
That's true, we could produce absolute keyframe values from the relative ones on animation start, then interpolate between those, instead of resolving relative values every frame. Good idea!

Reply:
I'm not sure taking a snapshot of the starting value is a good idea. I wouldn't expect that from this API design, I'd like to support things like implementing a follow behavior as a looping animation. Let's discuss.

Reply:
Let's leave it for now. We can improve it in an optimization pass.

Reply:
Just to expand on my comment, if this method instead returned a closure that returns the desired field, it could be cached in the keyframe. That way you'd avoid having to re-resolve the referenced field each frame.

Reply:
Are we talking about the same thing? I don't think you could implement a follow behavior with this. This is where you add the interpolated keyframe value to the animating object's original property value, as it was at the start of the animation.